solidity/core/contracts/validator-manager/OutboxValidatorManager.sol

```diff
@@ -49,25 +82,150 @@ contract OutboxValidatorManager is MultisigValidatorManager {
     // ============ External Functions ============
 
     /**
-     * @notice Determines if a quorum of validators have signed an improper checkpoint,
+     * @notice Determines if a quorum of validators have signed an invalid checkpoint,
+     * failing the Outbox if so.
+     * A checkpoint is invalid if it commits to anything other than contiguous non-empty leaves
+     * from leaf index zero to the leaf index of the checkpoint.
+     * @dev Invalid checkpoints signed by individual validators are not handled to prevent
+     * a single byzantine validator from failing the Outbox.
+     * @param _outbox The outbox.
+     * @param _signedRoot The root of the signed checkpoint.
+     * @param _signedIndex The index of the signed checkpoint.
+     * @param _signatures Signatures over the checkpoint to be checked for a validator
+     * quorum. Must be sorted in ascending order by signer address.
+     * @param _invalidLeaf The differing element in the fraudulent tree.
+     * @param _invalidProof Proof of inclusion of `_fraudulentLeaf`.
+     * @param _invalidIndex The index of the disputed leaf.
+     * @return True iff invalidity was proved.
+     */
+    function invalidCheckpoint(
+        IOutbox _outbox,
+        bytes32 _signedRoot,
+        uint256 _signedIndex,
+        bytes[] calldata _signatures,
+        bytes32 _invalidLeaf,
+        bytes32[32] calldata _invalidProof,
+        uint256 _invalidIndex
+    ) external returns (bool) {
+        require(isQuorum(_signedRoot, _signedIndex, _signatures), "!quorum");
+
+        bytes32 _invalidRoot = MerkleLib.branchRoot(
+            _invalidLeaf,
+            _invalidProof,
+            _invalidIndex
+        );
+        require(_invalidRoot == _signedRoot, "!root");
+
+        bool _invalidIfNonEmptyLeaf = _invalidIndex > _signedIndex;
+        bool _nonEmptyLeaf = _invalidLeaf != 0;
+        require(_invalidIfNonEmptyLeaf == _nonEmptyLeaf, "!invalid");
```

Comment:
This is clever but tbh I found this a little hard to understand. Wonder if there's a more readable way of articulating the same thing?

I'm also wondering:
1. Is the case where `_invalidIndex <= _signedIndex && _invalidLeaf == 0` that important to cover? Presumably someone would only want to do that to censor a leaf, but they could always just set the leaf to random bytes with the same effect without being slashed
2. Could the case where `_invalidIndex > _signedIndex && _invalidLeaf != 0` be entirely avoided by enforcing on the Inbox that the `index` of a message is <= the `index` of the checkpoint it's proving against?

Reply:
1. Agreed that it's not important to cover, and changing fraud proofs to utilize internal nodes will get us this for free.
2. Yeah excellent point, I believe this will allow us to get rid of `invalidCheckpoint` entirely